### PR TITLE
Human vs Human

### PR DESCRIPTION
**Pull request by andrewMacmurray**
+ adds human vs human game

### CODE REVIEW

File: lib/console/board_renderer.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class BoardRenderer
+  def render(board)
+    tile_strings = board.tiles.map { |tile| tile.to_s }
```
Comment (laurentqro) [inline lib/console/board_renderer.rb]: Ruby gives you this idiomatic shorthand for that kind of stuff: `board.tiles.map(&:to_s)`

https://www.brianstorti.com/understanding-ruby-idiom-map-with-symbol/

File: Gemfile.lock [hunk 2]
```diff
@@ -15,11 +48,16 @@ GEM
...
     rspec-support (3.7.1)
+    ruby_dep (1.5.0)
+    shellany (0.0.1)
+    thor (0.20.0)
 
 PLATFORMS
   ruby
 
 DEPENDENCIES
+  guard (~> 2.14)
```
Comment (fabientownsend) [inline Gemfile.lock:59]: Is this dependency necessary for your implementation? If it's not the case, it can be good to make it part of another PR

File: bin/console [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
+require "console/console"
+require "core/players/players"
+require "core/players/player_symbol"
+require "core/board"
+require "core/game"
 
-# (If you use this, don't forget to add pry to your Gemfile!)
-# require "pry"
-# Pry.start
+console = Console.new(input: $stdin, output: $stdout)
```
Comment (onlyskin) [inline bin/console]: I'd be tempted to hide these as defaults inside the class constructor with something like:
`@input = options[:input] || $stdin`.

File: lib/console/console.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    @input = options[:input]
+    @output = options[:output]
+  end
+
+  def greet_user
+    print(messages.greet_user)
+  end
+
+  def show_options
+    messages.options.each { |option| print(option) }
```
Comment (onlyskin) [inline lib/console/console.rb:17]: Similar to another comment, I think you can replace this block with `&:print`

File: lib/console/console.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  def get_move
+    input.gets.chomp.to_r
+  end
+
+  def clear
+    print(`clear`)
+  end
+
+  private
+  attr_reader :board_renderer
```
Comment (onlyskin) [inline lib/console/console.rb]: I've more commonly seen all the `attr_reader`'s on one line:
`attr_reader :board_renderer, :messages, :output, :input`

File: lib/console/console.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  attr_reader :messages
+  attr_reader :output
+  attr_reader :input
+
+  def print(message)
+    output.puts(message)
+  end
+
+  def board_renderer        
+    BoardRenderer.new
```
Comment (onlyskin) [inline lib/console/console.rb]: There's a common pattern in ruby to do something like:
`@board_renderer ||= BoardRenderer.new` which is a one-liner for lazy instance variable initialization

File: lib/core/board.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  def is_terminal?
+    is_full? || has_won?(:X) || has_won?(:O)
+  end
+
+  def is_full?
+    tiles.all? { |t| !t.is_empty? }
+  end
+
+  def has_won?(player_symbol)
+    winning_combinations.any? { |c| has_winning_combination?(c, player_symbol) }
```
Comment (onlyskin) [inline lib/core/board.rb]: I think it's better not to use single letter variables for anything that's not a simple index. It's obvious what `c` is when you read the method properly, but for someone skimming to get a sense of what's going on it obscures the details and forces you to slow down. (There's quite a few examples of that in this file, and you actually use `c` to mean something different further down in `has_winning_combination?`)

File: lib/core/game.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    play_round
+  end
+
+  private
+  attr_reader :players
+  attr_reader :ui
+  attr_accessor :board
+
+  def play_round
+    while !board.is_terminal? do
```
Comment (onlyskin) [inline lib/core/game.rb]: The method name `is_terminal?` may be confusing, I thought you were referring to the console.

File: spec/console/board_renderer_spec.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    board = Board.new
+    board_renderer = BoardRenderer.new
+
+    expected_lines = [
+      "1 | 2 | 3",
+      "4 | 5 | 6",
+      "7 | 8 | 9"
+    ].join("\n")
+
+    actual_lines = board_renderer.render(board)
```
Comment (onlyskin) [inline spec/console/board_renderer_spec.rb]: I would put this above the `expected_lines` declaration in order to follow 'Arrange, Act, Assert'. I think declaring what is expected is part of the test assertion because if it were shorter you'd have inlined it as part of the assertion.

File: spec/core/game_spec.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    expect(board.is_full?).to be(true)
+  end
+
+  def ui_spy(move_sequence)
+    ui = spy()
+    allow(ui).to receive(:get_move) { move_sequence.shift }
+    allow(ui).to receive(:show_win)
+    allow(ui).to receive(:show_draw)
+
+    expect(ui).to receive(:greet_user).once
```
Comment (onlyskin) [inline spec/core/game_spec.rb:52]: It's kindof unusual that these `expects` are inside a helper method – are they actually getting run?
It might give better coverage to have a third test case where you make these assertions.

File: lib/console/board_renderer.rb (hunk 1 above)
Comment (andrewMacmurray) [reply lib/console/board_renderer.rb]: nice!

File: lib/console/console.rb (hunk 6 above)
Comment (andrewMacmurray) [reply lib/console/console.rb]: all these one liners are ✨

File: lib/core/game.rb (hunk 8 above)
Comment (andrewMacmurray) [reply lib/core/game.rb]: I'd never thought of that, good point!
